Add WKWebsiteDataStore SPI to restore Local Storage data

RupinMittal opened the pull request:
#### b4877c19e09871c820b1b3aa103cff506defec42
<pre>
Add WKWebsiteDataStore SPI to restore Local Storage data
<a href="https://bugs.webkit.org/show_bug.cgi?id=284091">https://bugs.webkit.org/show_bug.cgi?id=284091</a>
<a href="https://rdar.apple.com/141009843">rdar://141009843</a>

Reviewed by Sihui Liu, Chris Dumez, and Brady Eidson.

Some clients may want to restore the local storage of ephemeral
data stores. (For example, Safari restoring the local storage of
private browsing tabs).

The restoration process will have two parts:
1. Client fetches the local storage data from WebKit and holds onto it.
2. Client gives back the data to WebKit to restore it.

The first part (fetch) has already landed: <a href="https://commits.webkit.org/287349@main.">https://commits.webkit.org/287349@main.</a>

This patch adds an SPI (which will eventually go through API
review) that does the second part of the restoration (restore).

The SPI works for both ephemeral and persistent data stores,
so there is an API test for each.

The SPI also works for both first party storage and for third party
storage, so there are API tests for each.

* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::restoreLocalStorage):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/storage/LocalStorageManager.cpp:
(WebKit::LocalStorageManager::ensureLocalStorageArea):
(WebKit::LocalStorageManager::connectToLocalStorageArea):
(WebKit::LocalStorageManager::ensureTransientLocalStorageArea):
(WebKit::LocalStorageManager::connectToTransientLocalStorageArea):
(WebKit::LocalStorageManager::populateStorageArea):
(WebKit::LocalStorageManager::protectedLocalStorageArea const): Deleted.
* Source/WebKit/NetworkProcess/storage/LocalStorageManager.h:
* Source/WebKit/NetworkProcess/storage/MemoryStorageArea.cpp:
(WebKit::MemoryStorageArea::setItem):
* Source/WebKit/NetworkProcess/storage/MemoryStorageArea.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::fetchLocalStorage):
(WebKit::NetworkStorageManager::restoreLocalStorage):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.cpp:
(WebKit::SQLiteStorageArea::setItem):
* Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.h:
* Source/WebKit/NetworkProcess/storage/StorageAreaBase.h:
* Source/WebKit/Shared/WebsiteData/WebsiteDataType.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(filterSupportedTypes):
(-[WKWebsiteDataStore _fetchDataOfTypes:completionHandler:]):
(-[WKWebsiteDataStore _restoreData:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::fetchLocalStorage):
(WebKit::NetworkProcessProxy::restoreLocalStorage):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::fetchLocalStorage):
(WebKit::WebsiteDataStore::restoreLocalStorage):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreLocalStorage.mm:
(testRestoreLocalStorage):
(TEST(WebKit, RestoreLocalStorageFromPersistentDataStore)):
(TEST(WebKit, RestoreLocalStorageFromEphemeralDataStore)):
(item):
(RestoreLocalStorageFromPersistentDataStoreThirdPartyIFrame)):
(RestoreLocalStorageFromEphemeralDataStoreThirdPartyIFrame)):
(testFetchLocalStorage): Deleted.
(TEST(WebKit, FetchLocalStorageFromPersistentDataStore)): Deleted.
(TEST(WebKit, FetchLocalStorageFromEphemeralDataStore)): Deleted.
(setItem): Deleted.
(FetchLocalStorageFromPersistentDataStoreThirdPartyIFrame)): Deleted.
(FetchLocalStorageFromEphemeralDataStoreThirdPartyIFrame)): Deleted.

Canonical link: <a href="https://commits.webkit.org/287641@main">https://commits.webkit.org/287641@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1ca11eb6cfbf20f97932907d609cb5eb4ab4d71d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80354 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59360 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33864 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84875 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31336 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82465 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68422 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7661 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62805 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20614 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83423 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52884 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73164 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43108 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/50204 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27320 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29795 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/71342 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27836 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86309 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7579 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/5362 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71087 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7754 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69003 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70327 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14320 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13268 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12434 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7541 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/13061 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7380 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10900 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9186 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
```diff
@@ -156,6 +156,7 @@ typedef NS_ENUM(uint8_t, _WKRestrictedOpenerType) {
 - (void)_runningOrTerminatingServiceWorkerCountForTesting:(void(^)(NSUInteger))completionHandler WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
 
 - (void)_fetchDataOfTypes:(NSSet<NSString *> *)dataTypes completionHandler:(WK_SWIFT_UI_ACTOR void (^)(NSData *))completionHandler;
+- (void)_restoreDataOfTypes:(NSSet<NSString *> *)dataTypes data:(NSData *)data;
```
this should have availability annotations, as well as the one above it

szewai commented on Source/WebKit/NetworkProcess/storage/LocalStorageType.h:
```diff
@@ -0,0 +1,35 @@
+
+namespace WebKit {
+
+enum class LocalStorageType : bool {
```
The type name is wrong: it should be memory / disk (or ephemeral / persistent).
You can tell if it's first-party data by looking at `ClientOrigin`.

RupinMittal replied:
Oh I didn't realize we can tell if it's first-party data by looking at ClientOrigin. How can I do that? 

Because in that case, we won't need to store this enum in the HashMap and pipe it all the way through. We can just decide where to put the data at point of storage in LocalStorageManager based on the ClientOrigin.

szewai replied:
If topOrigin equals clientOrigin, it's first-party.
Technically you will need the information to decide whether to create `m_transientStorageArea` or `m_localStorageArea` in `LocalStorageManager` for restoration.

RupinMittal replied:
Sweet! Let me try that. Thank you!

Right, since we're storing the Client Origin, from which we can determine first party vs third party, we'll no longer need  to store the enum in the data we fetch since it would be redundant.

RupinMittal commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreLocalStorage.mm:
```diff
@@ -60,22 +62,58 @@ static void testFetchLocalStorage(RetainPtr<WKWebsiteDataStore> websiteDataStore
+
+    [[[newWebView configuration] websiteDataStore] _restoreDataOfTypes:set.get() data:localStorageData.get()];
+
+    WTF::sleep(1_s);
```
I'm certain that this is wrong. Since `restoreDataOfTypes` will use async IPC but doesn't need a completion handler, how else can I wait for call the finish? It seems wrong to add a completion handler only for testing purposes.

rr-codes replied:
I personally think the WKDataStore SPI itself _should_ have a completion handler, because it is asynchronous in actuality.

RupinMittal replied:
That's a good point. I can't think of any other way the client can know when the function is finished. I've added a completion handler.

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
```diff
@@ -155,7 +155,8 @@ typedef NS_ENUM(uint8_t, _WKRestrictedOpenerType) {
 
-- (void)_fetchDataOfTypes:(NSSet<NSString *> *)dataTypes completionHandler:(WK_SWIFT_UI_ACTOR void (^)(NSData *))completionHandler;
+- (void)_fetchDataOfTypes:(NSSet<NSString *> *)dataTypes completionHandler:(WK_SWIFT_UI_ACTOR void (^)(NSData *))completionHandler WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
+- (void)_restoreDataOfTypes:(NSSet<NSString *> *)dataTypes data:(NSData *)data WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
```
(needs visionos too)

rr-codes commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreLocalStorage.mm:
```diff
@@ -60,22 +62,58 @@ static void testFetchLocalStorage(RetainPtr<WKWebsiteDataStore> websiteDataStore
+
+    WTF::sleep(1_s);
+
+    auto newNavigationDelegate = adoptNS([[TestNavigationDelegate alloc] init]);
```
RetainPtr instead of auto

rr-codes commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreLocalStorage.mm:
```diff
@@ -60,22 +62,58 @@ static void testFetchLocalStorage(RetainPtr<WKWebsiteDataStore> websiteDataStore
+
+    // Start with a clean slate data store.
+    __block bool doneRemoving = false;
+    [[WKWebsiteDataStore defaultDataStore] removeDataOfTypes:[WKWebsiteDataStore allWebsiteDataTypes] modifiedSince:[NSDate distantPast] completionHandler:^() {
```
no need for ()

RupinMittal replied:
Fixed!

rr-codes commented on Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.cpp:
```diff
@@ -379,6 +378,22 @@ Expected<void, StorageError> SQLiteStorageArea::setItem(IPC::Connection::UniqueI
+
+Expected<void, StorageError> SQLiteStorageArea::setItem(const String& key, const String& value)
+{
+    String oldValue;
```
what's the point of `oldValue` here?

RupinMittal replied:
It's not used here but the other overload of `setItem()` needs to pass in `oldValue` since it dispatches a message with that value. Since I wanted to move much of the logic of setting the item to another function to avoid code duplication, I needed to pass in that parameter here as well.

RupinMittal commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreLocalStorage.mm:
```diff
@@ -50,32 +60,76 @@ static void testFetchLocalStorage(RetainPtr<WKWebsiteDataStore> websiteDataStore
+    doneEvaluatingJavaScript = false;
+    [newWebView evaluateJavaScript:@"localStorage.getItem('key');" completionHandler:^(id value, NSError *error) {
+        EXPECT_NULL(error);
+        // This is currently failing (value is null).
```
Progress report: I think the overall structure of the patch mostly there. But it's currently failing :(. Will debug more tomorrow.

RupinMittal replied:
It works now :)

beidson commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
```diff
@@ -155,7 +155,8 @@ typedef NS_ENUM(uint8_t, _WKRestrictedOpenerType) {
 
-- (void)_fetchDataOfTypes:(NSSet<NSString *> *)dataTypes completionHandler:(WK_SWIFT_UI_ACTOR void (^)(NSData *))completionHandler;
+- (void)_fetchDataOfTypes:(NSSet<NSString *> *)dataTypes completionHandler:(WK_SWIFT_UI_ACTOR void(^)(NSData *))completionHandler WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA));
+- (void)_restoreDataOfTypes:(NSSet<NSString *> *)dataTypes data:(NSData *)data completionHandler:(WK_SWIFT_UI_ACTOR void(^)(void))completionHandler WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA));
```
I think the types that are fetched are encoded in the `_fetchDataOfTypes` step above, and the restore step should just ingest them all.
To many weird edge cases arise from allowing the client to choose a different set of types on the restore step, and I'm not sure it's useful to do.

RupinMittal replied:
That's a good point. I've updated the fetch step to encode the data type and the restore step to just ingest them.

beidson commented on Source/WebKit/NetworkProcess/storage/LocalStorageManager.h:
```diff
@@ -66,12 +66,15 @@ class LocalStorageManager {
 
     HashMap<String, String> fetchStorageMap() const;
+    void setLocalStorageMap(HashMap<String, String>, const WebCore::ClientOrigin&, Ref<WorkQueue>&&);
+    void setTransientLocalStorageMap(HashMap<String, String>, const WebCore::ClientOrigin&);
```
HashMap<String, String>&& ?

beidson replied:
(Or, if WTFMove semantics don't work for some reason, at the very least `const HashMap<String, String>&`

RupinMittal replied:
Fixed!

beidson commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -1447,6 +1449,37 @@ - (void)_fetchDataOfTypes:(NSSet<NSString *> *)dataTypes completionHandler:(WK_S
     completionHandler(nil);
 }
 
+- (void)_restoreData:(NSData *)data completionHandler:(void(^)(void))completionHandler
```
This method works fine today with one type.
But in the future - if two optional types exist - it will have to be robust against any of the types missing (e.g. not early return as error handling)

Not critical to account for with code now, but maybe a comment documenting for the future.

RupinMittal replied:
That's a good point. I've updated the code to be better built for adding more types later as Chris recommended.

beidson reviewed (approved):
r+ with some small changes

cdumez commented on Source/WebKit/NetworkProcess/storage/LocalStorageManager.cpp:
```diff
@@ -254,4 +261,18 @@ HashMap<String, String> LocalStorageManager::fetchStorageMap() const
     return { };
 }
 
+void LocalStorageManager::setLocalStorageMap(HashMap<String, String> storageMap, const WebCore::ClientOrigin& origin, Ref<WorkQueue>&& workQueue)
```
`HashMap<String, String>&&`

cdumez commented on Source/WebKit/NetworkProcess/storage/LocalStorageManager.cpp:
```diff
@@ -254,4 +261,18 @@ HashMap<String, String> LocalStorageManager::fetchStorageMap() const
+        localStorageArea->setItem(key, value);
+}
+
+void LocalStorageManager::setTransientLocalStorageMap(HashMap<String, String> storageMap, const WebCore::ClientOrigin& origin)
```
`HashMap<String, String>&&`

cdumez commented on Source/WebKit/NetworkProcess/storage/LocalStorageManager.cpp:
```diff
@@ -185,7 +185,7 @@ void LocalStorageManager::connectionClosedForTransientStorageArea(IPC::Connectio
 }
 
-StorageAreaIdentifier LocalStorageManager::connectToLocalStorageArea(IPC::Connection::UniqueID connection, StorageAreaMapIdentifier sourceIdentifier, const WebCore::ClientOrigin& origin, Ref<WorkQueue>&& workQueue)
+StorageAreaBase* LocalStorageManager::ensureLocalStorageArea(const WebCore::ClientOrigin& origin, Ref<WorkQueue>&& workQueue)
```
Why is this returning a raw pointer and not a C++ reference? Seems like it can never return null?

cdumez commented on Source/WebKit/NetworkProcess/storage/LocalStorageManager.cpp:
```diff
@@ -198,17 +198,18 @@ StorageAreaIdentifier LocalStorageManager::connectToLocalStorageArea(IPC::Connec
 }
 
-StorageAreaIdentifier LocalStorageManager::connectToTransientLocalStorageArea(IPC::Connection::UniqueID connection, StorageAreaMapIdentifier sourceIdentifier, const WebCore::ClientOrigin& origin)
+MemoryStorageArea* LocalStorageManager::ensureTransientLocalStorageArea(const WebCore::ClientOrigin& origin)
```
Why is this returning a raw pointer and not a C++ reference? Seems like it can never return null?

cdumez commented on Source/WebKit/NetworkProcess/storage/LocalStorageManager.cpp:
```diff
@@ -217,6 +218,12 @@ StorageAreaIdentifier LocalStorageManager::connectToTransientLocalStorageArea(IP
+
+StorageAreaIdentifier LocalStorageManager::connectToTransientLocalStorageArea(IPC::Connection::UniqueID connection, StorageAreaMapIdentifier sourceIdentifier, const WebCore::ClientOrigin& origin)
+{
+    RefPtr transientStorageArea = ensureTransientLocalStorageArea(origin);
```
`Ref`

cdumez commented on Source/WebKit/NetworkProcess/storage/LocalStorageManager.cpp:
```diff
@@ -198,17 +198,18 @@ StorageAreaIdentifier LocalStorageManager::connectToLocalStorageArea(IPC::Connec
+StorageAreaIdentifier LocalStorageManager::connectToLocalStorageArea(IPC::Connection::UniqueID connection, StorageAreaMapIdentifier sourceIdentifier, const WebCore::ClientOrigin& origin, Ref<WorkQueue>&& workQueue)
 {
-    return m_localStorageArea;
+    RefPtr localStorageArea = ensureLocalStorageArea(origin, WTFMove(workQueue));
```
`Ref`

cdumez commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -1341,6 +1341,29 @@ void NetworkStorageManager::fetchLocalStorage(CompletionHandler<void(HashMap<Web
+                localStorageManager.setTransientLocalStorageMap(WTFMove(storageMap), clientOrigin);
+        }
+
+        RunLoop::protectedMain()->dispatch([completionHandler = WTFMove(completionHandler)] mutable {
```
I believe you need to capture `protectedThis = WTFMove(protectedThis)` here too. We need to make sure to deref on the main thread so the NetworkStorageManager gets destroyed always on the main thread. Some of its data members are Strings so the object cannot be destroyed on random threads.

RupinMittal replied:
I've fixed this and that does make sense--but I'm curious--some of the other functions in this file don't do that (for example `NetworkStorageManager::requestSpace`). How come we don't need to capture `protectedThis` there?

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -1447,6 +1449,37 @@ - (void)_fetchDataOfTypes:(NSSet<NSString *> *)dataTypes completionHandler:(WK_S
+        return;
+    }
+
+    if (dataTypes->contains(WKWebsiteDataTypeLocalStorage)) {
```
This serialization / deserialization format is not right. If we supported more than 1 type, this wouldn't work.

Upon decoding, you wouldn't know the decoding order of different types.

RupinMittal replied:
Asked about this on Slack.

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -1429,14 +1430,15 @@ - (void)_runningOrTerminatingServiceWorkerCountForTesting:(void(^)(NSUInteger))c
 
             WTF::Persistence::Encoder encoder;
             encoder << currentLocalStorageSerializationVersion;
+            encoder << HashSet<String> { WKWebsiteDataTypeLocalStorage };
```
I don't think we should encode a String since this is very inefficient. Ideally, this would be a enum.

Also, we don't think we want to encode a HashSet. A Vector would be better so we have strict ordering.

RupinMittal replied:
Updated to use enum. I've asked a question about the rest on Slack.

cdumez commented on Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
```diff
@@ -2003,13 +2003,20 @@ void NetworkProcessProxy::setEmulatedConditions(PAL::SessionID sessionID, std::o
 {
-    if (!canSendMessage()) {
+    if (!canSendMessage())
         completionHandler({ });
```
This needs a return statement after.

szewai replied:
I think Rupin misunderstood my [comment](https://github.com/WebKit/WebKit/pull/37055#discussion_r1868472226): I meant the check is not needed (because `AuxiliaryProcessProxy::sendMessage` checks process state there), not `return` is not needed

RupinMittal replied:
Ah yes, I did misunderstand. That makes sense. I've removed the check.

cdumez commented on Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
```diff
@@ -2003,13 +2003,20 @@ void NetworkProcessProxy::setEmulatedConditions(PAL::SessionID sessionID, std::o
+void NetworkProcessProxy::restoreLocalStorage(PAL::SessionID sessionID, HashMap<WebCore::ClientOrigin, HashMap<String, String>>&& localStorage, CompletionHandler<void()>&& completionHandler)
+{
+    if (!canSendMessage())
+        completionHandler();
```
This needs a return statement after.

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
```diff
@@ -155,7 +155,8 @@ typedef NS_ENUM(uint8_t, _WKRestrictedOpenerType) {
 - (void)_runningOrTerminatingServiceWorkerCountForTesting:(void(^)(NSUInteger))completionHandler WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
 
-- (void)_fetchDataOfTypes:(NSSet<NSString *> *)dataTypes completionHandler:(WK_SWIFT_UI_ACTOR void (^)(NSData *))completionHandler;
+- (void)_fetchDataOfTypes:(NSSet<NSString *> *)dataTypes completionHandler:(WK_SWIFT_UI_ACTOR void(^)(NSData *))completionHandler WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA));
```
I think we need a comment to explain which data types are supported.

cdumez reviewed (changes requested):
Needs some work.

szewai commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -1341,6 +1341,29 @@ void NetworkStorageManager::fetchLocalStorage(CompletionHandler<void(HashMap<Web
+            if (clientOrigin.topOrigin == clientOrigin.clientOrigin)
+                localStorageManager.setLocalStorageMap(WTFMove(storageMap), clientOrigin, protectedWorkQueue());
+            else
+                localStorageManager.setTransientLocalStorageMap(WTFMove(storageMap), clientOrigin);
```
`LocalStorageManager` should just have one `populateLocalStorageArea()`, and whether to create a TransientLocalStorageArea or LocalStorageArea should be moved to there.

szewai commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -1341,6 +1341,29 @@ void NetworkStorageManager::fetchLocalStorage(CompletionHandler<void(HashMap<Web
+    ASSERT(RunLoop::isMain());
+    ASSERT(!m_closed);
+
+    protectedWorkQueue()->dispatch([this, protectedThis = Ref { *this }, localStorageMap = crossThreadCopy(localStorageMap), completionHandler = WTFMove(completionHandler)]() mutable {
```
`localStorageMap = crossThreadCopy(WTFMove(localStorageMap))`

szewai commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -1341,6 +1341,29 @@ void NetworkStorageManager::fetchLocalStorage(CompletionHandler<void(HashMap<Web
+        assertIsCurrent(workQueue());
+
+        for (auto& [clientOrigin, storageMap] : localStorageMap) {
+            auto& localStorageManager = originStorageManager(clientOrigin, ShouldWriteOriginFile::No).localStorageManager(*m_storageAreaRegistry);
```
This needs to be `ShouldWriteOriginFile::YES`. This is a write operation, and origin file should be created when origin has storage.

szewai commented on Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.cpp:
```diff
@@ -341,7 +341,7 @@ HashMap<String, String> SQLiteStorageArea::allItems()
 }
 
-Expected<void, StorageError> SQLiteStorageArea::setItem(IPC::Connection::UniqueID connection, StorageAreaImplIdentifier storageAreaImplID, String&& key, String&& value, const String& urlString)
+Expected<void, StorageError> SQLiteStorageArea::setItem(const String& key, const String& value, String& oldValue)
```
Why not just make `IPC::Connection::UniqueID` optional, and skip event dispatch if connectionID is null? Why do we need so many `setItem` functions?

RupinMittal replied:
Fixed!

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreLocalStorage.mm:
```diff
@@ -119,8 +174,46 @@ async function setItem()
 )TESTRESOURCE"_s;
 
-static void testFetchLocalStorageThirdPartyIFrame(RetainPtr<WKWebsiteDataStore> websiteDataStore)
+static NSString *getMainFrameString = @"<script> \
```
How is `getMainFrameString` different from `setMainFrameString`? Can we have one string?

RupinMittal replied:
Oops, that's a remnant from my attempt to host each iFrame on a different port. Fixed!

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreLocalStorage.mm:
```diff
@@ -138,31 +231,82 @@ static void testFetchLocalStorageThirdPartyIFrame(RetainPtr<WKWebsiteDataStore>
+    // Kill the network and web processes and the server.
+    [websiteDataStore _terminateNetworkProcess];
+    [webView _killWebContentProcessAndResetState];
+    server.cancel();
```
Why do we need to kill the server and processes? Data is removed after `removeDataOfTypes` (if you want to validate you could `localStorage.setItem()` in WebView with `evaluateJavaScript` and `getItem()` after `removeDataOfTypes` call).

You could just reload the web view after `_restoreData` and check what message is received.

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreLocalStorage.mm:
```diff
@@ -119,8 +174,46 @@ async function setItem()
+    </script> \
+    <iframe src='https://127.0.0.1:9091/'>";
+
+static constexpr auto getFrameBytes = R"TESTRESOURCE(
```
I don't think you need to separate `getFrameBytes` and `setFrameBytes`. Just have one `frameBytes`, which invokes `getItem()` first, if item exists, post message with "found item [value]" and return; if not exists, `setItem()` and post message "set item".
You could tell whether item exists by the message.

RupinMittal replied:
That's a good point. Fixed!

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreLocalStorage.mm:
```diff
@@ -119,8 +174,46 @@ async function setItem()
+    } catch(err) {
+        postMessage('error: ' + err.name + ' - ' + err.message);
+    }
+}
```
Not quite sure why this function is `async` -- `getItem` is sync; and exception is unlikely to throw with existing setup (so `try... catch` is also unnecessary).

szewai reviewed (commented):
Is the SPI expected to merge data with existing data, or replace existing data? Your current implementation is merge. If you want to replace, you need to clear `StorageArea` before `setItem`. And you might want to add test for this.

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -1428,15 +1430,37 @@ - (void)_runningOrTerminatingServiceWorkerCountForTesting:(void(^)(NSUInteger))c
             encoder << currentLocalStorageSerializationVersion;
-            encoder << localStorage;
+            if (!localStorage.isEmpty()) {
+                encoder << WKWebsiteDataTypeSet { WKWebsiteDataType::WKWebsiteDataTypeLocalStorage };
```
The set should be encoded as a pre-step and include all the types included in `dataTypes` and that we support restoration for.

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -1428,15 +1430,37 @@ - (void)_runningOrTerminatingServiceWorkerCountForTesting:(void(^)(NSUInteger))c
+                encoder << WKWebsiteDataTypeSet { WKWebsiteDataType::WKWebsiteDataTypeLocalStorage };
+                encoder << localStorage;
+            } else
+                encoder << WKWebsiteDataTypeSet { };
```
Again should be a pre-step, before we check `if ([dataTypes containsObject:WKWebsiteDataTypeLocalStorage]) {`

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -1446,6 +1470,37 @@ - (void)_fetchDataOfTypes:(NSSet<NSString *> *)dataTypes completionHandler:(WK_S
+        completionHandler();
+        return;
+    }
+
```
We should construct a `CallbackAggregator` here from the completion handler.

RupinMittal replied:
Added!

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -1446,6 +1470,37 @@ - (void)_fetchDataOfTypes:(NSSet<NSString *> *)dataTypes completionHandler:(WK_S
+        if (!localStorage)
+            return;
+
+        _websiteDataStore->restoreLocalStorage(WTFMove(*localStorage), WTFMove(completionHandler));
```
This should pass a lambda that captures the CallbackAggregator

RupinMittal replied:
Added!

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -1446,6 +1470,37 @@ - (void)_fetchDataOfTypes:(NSSet<NSString *> *)dataTypes completionHandler:(WK_S
+            return;
+
+        _websiteDataStore->restoreLocalStorage(WTFMove(*localStorage), WTFMove(completionHandler));
+        return;
```
We should not early return here.

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -1428,22 +1431,85 @@ - (void)_runningOrTerminatingServiceWorkerCountForTesting:(void(^)(NSUInteger))c
-            encoder << currentLocalStorageSerializationVersion;
-            encoder << localStorage;
-            completionHandler(toNSData(encoder.span()).get());
+        for (size_t i = 0; i < dataTypesToEncode.size() ; i++) {
```
Please use a modern loop:
```
for (auto& dataTypeToEncode : dataTypesToEncode) {
```

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -1428,22 +1431,85 @@ - (void)_runningOrTerminatingServiceWorkerCountForTesting:(void(^)(NSUInteger))c
+        completionHandler();
+    });
+
+    for (size_t i = 0; i < encodedDataTypes->size() ; i++) {
```
```
for (auto& encodedDataType : *encodedDataTypes) {
```

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -1428,22 +1431,85 @@ - (void)_runningOrTerminatingServiceWorkerCountForTesting:(void(^)(NSUInteger))c
+    });
+
+    for (size_t i = 0; i < encodedDataTypes->size() ; i++) {
+        switch (encodedDataTypes->at(i)) {
```
```
switch (encodedDataType) {
```

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
```diff
@@ -155,7 +155,9 @@ typedef NS_ENUM(uint8_t, _WKRestrictedOpenerType) {
-- (void)_fetchDataOfTypes:(NSSet<NSString *> *)dataTypes completionHandler:(WK_SWIFT_UI_ACTOR void (^)(NSData *))completionHandler;
+// Supported data types are: WKWebsiteDataTypeLocalStorage.
+- (void)_fetchDataOfTypes:(NSSet<NSString *> *)dataTypes completionHandler:(WK_SWIFT_UI_ACTOR void(^)(NSData *))completionHandler WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA));
+- (void)_restoreData:(NSData *)data completionHandler:(WK_SWIFT_UI_ACTOR void(^)(void))completionHandler WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA));
```
Seems like the completionHandler should take a BOOL that indicates whether or not the operation was successful.

RupinMittal replied:
Added!

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -1428,22 +1431,85 @@ - (void)_runningOrTerminatingServiceWorkerCountForTesting:(void(^)(NSUInteger))c
+    std::optional<unsigned> currentLocalStorageSerializationVersion;
+    decoder >> currentLocalStorageSerializationVersion;
+    if (!currentLocalStorageSerializationVersion) {
+        completionHandler();
```
completionHandler(NO);

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -1428,22 +1431,85 @@ - (void)_runningOrTerminatingServiceWorkerCountForTesting:(void(^)(NSUInteger))c
+    std::optional<Vector<WebKit::WebsiteDataType>> encodedDataTypes;
+    decoder >> encodedDataTypes;
+    if (!encodedDataTypes) {
+        completionHandler();
```
completionHandler(NO);

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -1428,22 +1431,90 @@ - (void)_runningOrTerminatingServiceWorkerCountForTesting:(void(^)(NSUInteger))c
+        case WebKit::WebsiteDataType::LocalStorage: {
+            std::optional<HashMap<WebCore::ClientOrigin, HashMap<String, String>>> localStorage;
+            decoder >> localStorage;
+            if (localStorage && !localStorage->isEmpty()) {
```
`*succeeded` needs to be set to `false` if `!localStorage`

cdumez reviewed (approved):
LGTM with 1 fix.

szewai commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -1428,22 +1431,96 @@ - (void)_runningOrTerminatingServiceWorkerCountForTesting:(void(^)(NSUInteger))c
+    }
+}
+
+- (void)_restoreData:(NSData *)data completionHandler:(void(^)(BOOL))completionHandler
```
When this becomes API, `CompletionHandler` should take `NSError *` so client knows the cause

RupinMittal replied:
Ok got it. So we'll need to make this change after API review is complete and decided upon?

szewai commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -1428,22 +1431,96 @@ - (void)_runningOrTerminatingServiceWorkerCountForTesting:(void(^)(NSUInteger))c
+                encoder << data->localStorage.value();
+                break;
+            default:
+                break;
```
```
ASSERT_NOT_REACHED();
break;
```

szewai commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -1428,22 +1431,96 @@ - (void)_runningOrTerminatingServiceWorkerCountForTesting:(void(^)(NSUInteger))c
+            break;
+        }
+        default:
+            break;
```
```
ASSERT_NOT_REACHED();
break;
```

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreLocalStorage.mm:
```diff
@@ -50,32 +61,73 @@ static void testFetchLocalStorage(RetainPtr<WKWebsiteDataStore> websiteDataStore
+    done = false;
+
+    // Create a new WebView.
+    RetainPtr newViewConfiguration = adoptNS([WKWebViewConfiguration new]);
```
You don't need to create new view configuration or new navigation delegate. You could reuse the existing ones for the new web view.

RupinMittal replied:
Ah you're right, that makes the test a bit simpler. Thank you!

RupinMittal committed:
Add WKWebsiteDataStore SPI to restore Local Storage data
https://bugs.webkit.org/show_bug.cgi?id=284091
rdar://141009843

Reviewed by Sihui Liu, Chris Dumez, and Brady Eidson.

Some clients may want to restore the local storage of ephemeral
data stores. (For example, Safari restoring the local storage of
private browsing tabs).

The restoration process will have two parts:
1. Client fetches the local storage data from WebKit and holds onto it.
2. Client gives back the data to WebKit to restore it.

The first part (fetch) has already landed: https://commits.webkit.org/287349@main.

This patch adds an SPI (which will eventually go through API
review) that does the second part of the restoration (restore).

The SPI works for both ephemeral and persistent data stores,
so there is an API test for each.

The SPI also works for both first party storage and for third party
storage, so there are API tests for each.

* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::restoreLocalStorage):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/storage/LocalStorageManager.cpp:
(WebKit::LocalStorageManager::ensureLocalStorageArea):
(WebKit::LocalStorageManager::connectToLocalStorageArea):
(WebKit::LocalStorageManager::ensureTransientLocalStorageArea):
(WebKit::LocalStorageManager::connectToTransientLocalStorageArea):
(WebKit::LocalStorageManager::populateStorageArea):
(WebKit::LocalStorageManager::protectedLocalStorageArea const): Deleted.
* Source/WebKit/NetworkProcess/storage/LocalStorageManager.h:
* Source/WebKit/NetworkProcess/storage/MemoryStorageArea.cpp:
(WebKit::MemoryStorageArea::setItem):
* Source/WebKit/NetworkProcess/storage/MemoryStorageArea.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::fetchLocalStorage):
(WebKit::NetworkStorageManager::restoreLocalStorage):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.cpp:
(WebKit::SQLiteStorageArea::setItem):
* Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.h:
* Source/WebKit/NetworkProcess/storage/StorageAreaBase.h:
* Source/WebKit/Shared/WebsiteData/WebsiteDataType.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(filterSupportedTypes):
(-[WKWebsiteDataStore _fetchDataOfTypes:completionHandler:]):
(-[WKWebsiteDataStore _restoreData:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::fetchLocalStorage):
(WebKit::NetworkProcessProxy::restoreLocalStorage):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::fetchLocalStorage):
(WebKit::WebsiteDataStore::restoreLocalStorage):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreLocalStorage.mm:
(testRestoreLocalStorage):
(TEST(WebKit, RestoreLocalStorageFromPersistentDataStore)):
(TEST(WebKit, RestoreLocalStorageFromEphemeralDataStore)):
(item):
(RestoreLocalStorageFromPersistentDataStoreThirdPartyIFrame)):
(RestoreLocalStorageFromEphemeralDataStoreThirdPartyIFrame)):
(testFetchLocalStorage): Deleted.
(TEST(WebKit, FetchLocalStorageFromPersistentDataStore)): Deleted.
(TEST(WebKit, FetchLocalStorageFromEphemeralDataStore)): Deleted.
(setItem): Deleted.
(FetchLocalStorageFromPersistentDataStoreThirdPartyIFrame)): Deleted.
(FetchLocalStorageFromEphemeralDataStoreThirdPartyIFrame)): Deleted.

Canonical link: https://commits.webkit.org/287641@main
